### PR TITLE
At 33

### PR DESCRIPTION
**Pull request by OllieVanD**
Added Student file uploader, with a maximum of 50 lines of entries. Added Jasmine Unit testing for the class, also had a feature test for the overall class functionality. Used Cypress testing also for another form of feature test. 

Refactored some of the code in file uploader and changed some of the Jasmine tests for other classes to match the changes.

### CODE REVIEW

File: spec/unit/file-uploader/dateCheck.spec.js [hunk 1]
```diff
@@ -1,15 +1,26 @@
...
-    });
-    it('shows errors for values after current date', () => {
-        const errors = dateCheck('08/17/2098 15:45', '08/17/2099 15:45', 5);
-        expect(errors).toEqual([
-            'Due date: 08/17/2098 15:45 is invalid, on line 5',
-            'Submission date: 08/17/2099 15:45 is invalid, on line 5'
-        ]);
-    });
-});
+  dateMockObject = {}
```
Comment (dearshrewdwit) [inline spec/unit/file-uploader/dateCheck.spec.js:4]: declare using `let`

File: spec/feature/studentChecker.spec.js [hunk 2]
```diff
@@ -0,0 +1,42 @@
+const StudentChecker = require('../../src/fileUpload/students.js')
+
+describe('feature tests for student checker', () => {
+  let StudentMockData = []
```
Comment (dearshrewdwit) [inline spec/feature/studentChecker.spec.js:4]: vars should not be capitalised

File: src/fileUpload/moduleChallenge.js [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+class ModuleChallengeChecker {
+  constructor () {
+    this.errors = []
+  }
+
+  async check (data) {
+    const students = await this.findAllStudents()
+
+    const arr = students.map(student => student.id)
```
Comment (dearshrewdwit) [inline src/fileUpload/moduleChallenge.js:17]: Naming: `studentIds`

File: src/fileUpload/fileUpload.js [hunk 4]
```diff
@@ -0,0 +1,82 @@
+const { SelfAssessment, Student, ModuleChallenge } = require('../../models')
+const SelfAssessmentChecker = require('./selfAssessment.js')
+const ModuleChallengeChecker = require('./moduleChallenge.js')
+const StudentChecker = require('./students.js')
+const headerChecker = require('./headerCheck.js')
+
+class FileUploader {
+  fileTypes = {
```
Comment (dearshrewdwit) [inline src/fileUpload/fileUpload.js:8]: declare using `const` - and move outside the class

**Review by dearshrewdwit — CHANGES_REQUESTED** [review]
Good. A few unresolved conversations to go 👍

**Review by Neokrieger — APPROVED** [review]
Should be good now(I hope)

File: src/fileUpload/fileUpload.js [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+        
+        this.history = table
+        
+        this.table = this.dbCheck(table)
+        
+        this.data = this.fileConvertor(csvFile)
+
+        this.validFile = headerChecker(this.headers, table)
+
+        if (typeof this.validFile === 'string') {
```
[the remark this answers is not in the document]
Comment (dearshrewdwit) [reply src/fileUpload/fileUpload.js:30]: Conversations to resolve @Neokrieger @OllieVanD

File: spec/feature/fileUploader.spec.js [hunk 6]
```diff
@@ -0,0 +1,28 @@
+const FileUploader = require('../../src/fileUpload/fileUpload.js')
```
[the remark this answers is not in the document]
Comment (dearshrewdwit) [reply spec/feature/fileUploader.spec.js:1]: Conversations to resolve @Neokrieger @OllieVanD

File: src/fileUpload/fileUpload.js [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+            }
+            arrayObj.push(obj)
+        }
+        return arrayObj
+    }   
+
+  async addToDatabase () {
+    if (this.errors.length === 0) {
+      await this.table.table.bulkCreate(this.data, {
+        returning: true
```
[the remark this answers is not in the document]
Comment (dearshrewdwit) [reply src/fileUpload/fileUpload.js:74]: Conversations to resolve @Neokrieger @OllieVanD

File: spec/feature/fileUploader.spec.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (OllieVanD) [reply spec/feature/fileUploader.spec.js:1]: I was going to do all of this in the AT-80 branch - refactor all code and add relevant Jasmine tests. But I can all of the tests onto this branch.

[the remark this answers is not in the document]
Comment (dearshrewdwit) [reply spec/feature/fileUploader.spec.js:1]: The refactoring can be on a separate PR 👍 

The test for the `#process` method should be here though.

**Review by Neokrieger — APPROVED** [review]

File: src/fileUpload/moduleChallenge.js [hunk 8]
```diff
@@ -1,61 +1,71 @@
...
+      this.projectCheck(dataObject.challengeName, dataObject.language, counter)
+      this.errors = this.errors.concat(dateCheck(dataObject.dueDate, dataObject.submissionDate, counter))
+      this.scoreCheck(dataObject.studentScore, dataObject.coachScore, counter)
+    })
+    return this.errors
+      }
+    
+    async findAllStudents() {
+        return await Student.findAll({
+            attributes: ['id'],
```
[the remark this answers is not in the document]
Comment (OllieVanD) [reply src/fileUpload/moduleChallenge.js]: This just checks that the StudentId exists from the Students table

**Review by dearshrewdwit — APPROVED** [review]

**Review by Neokrieger — APPROVED** [review]

**Pull request merged**